### ISSUE TITLE
Memory Leak - Threads are leaking

### ISSUE THREAD

**Issue by aguyyala-disco** [xtream1101/s3-concat#11]
Hi @xtream1101 , Thanks a bunch for putting this together! We are using this for our project. We detected a memory leak in this library. The threads are being created every time we invoke a concat operation and they are not being cleaned up. This is exactly where the bug is https://github.com/xtream1101/s3-concat/blob/master/s3_concat/utils.py#L36. We are currently using ThreadPool to deal with this issue but it'd be great if you can fix it here so that we can tag along and use all great features of the library that will be coming in the future. Here's a quick glance at our fix.

```
from multiprocessing.pool import ThreadPool as Pool

def _thread_run(item, callback):
    for _ in range(3):
        # re try 3 times before giving up
        try:
            response = callback(item)
            return response
        except Exception:
            logger.exception("Retry failed batch of: {}".format(item))


def _threads(num_threads, data, callback):
    pool = Pool(num_threads)
    results = pool.starmap(_thread_run, [(item, callback) for item in data])
    pool.close()
    pool.join()
    return results
```

Please let us know if you need further details on this.

**Comment by theashworld** [comment]
hi @aguyyala-disco , is there a fork you have with this fix? I'm looking to use this library, and I suspect we will encounter this issue too.

**Comment by christophlingg** [comment]
There is another memory leak regarding the recreation of s3 clients. I used this patched class:

```python
class NotLeakingS3Concat(S3Concat):
    """
    Patched Version of S3Concat that avoids endless recreations of s3_clients   
    """
    def __init__(self, bucket, key, s3_client):
        # It is crazy not to call the base class constructor
        # but I want avoid the creation of a new s3 client
        self.bucket = bucket
        self.key = key
        self.min_file_size = None
        self.content_type = 'text/csv'
        self.all_files = []
        self.s3 = s3_client
```

But now I also run into the next memory leak described above. I created a fork requested a PR to incorporate both changes.

**Comment by christophlingg** [comment]
@aguyyala-disco would you mind explaining why the previous code caused a memory leak and yours doesn't?

**Comment by christophlingg** [comment]
With this snippet, you can monkey-patch s3_concat to avoid using threads their memory leak.

```python
import s3_concat

def _threads(nb_threads, data, callback):
    """
    monkey patched function that runs everything sequentially without using threads to avoid memory leaks
    see https://github.com/xtream1101/s3-concat/issues/11
    """
    return [
        callback(d)
        for d in data
    ]


s3_concat.multipart_upload_job._threads = _threads
```

**Comment by xtream1101** [comment]
This is now fixed in version `0.2.3`
